### ISSUE TITLE
Warn about using predicates only in negative position.

### ISSUE THREAD

**Issue by matzemathics** [knowsys/nemo#433]
Consider following program:

```prolog
a(12).
b(?x) :- a(?x), ~c(?x).
```

This seems fine, but does not satisfy the type checker:
```
error: Conflicting type declarations. Predicate "a" at position 1 has been inferred to have the conflicting types integer and any.
```
This is because `c` has the inferred type `any`, while `a` has the inferred type `integer`.
This type of error will only happen, if the predicate is only ever used in a negated position, so instead of adapting the type checker to consider negated usages of a predicate as type hints, I propose to warn about the absence of any positive occurrence in general, as that is most likely due to missing rules / mistyping etc.

**Comment by monsterkrampe** [comment]
This also happens for the following program:
```
a(12).
b(?x) :- a(?x), c(?x).
```
There is a comment about this here: https://github.com/knowsys/nemo/blob/6f3e3a880b2fff585447b1d60a1a8d7857cf03c9/nemo/src/program_analysis/type_inference/position_graph.rs#L303-L312

**Comment by mmarx** [comment]
Isn't this mostly a consequence of us not yet supporting type promotion? `any` and `integer` should not be incompatible at all, since `integer` is a subtype of `any`?

**Comment by monsterkrampe** [comment]
Also that, at some point `integer` and `any` should be compatible but right now the physical layer would not handle this correctly, so we reject this. This is what is encoded through the `partial_cmp` check.

**Comment by monsterkrampe** [comment]
I think we should leave this as is for now.
The current behavior is as follows:
Once all type inference and propagation has been done, all unassigned types are just set to `Any`.
Afterwards, types in bodies are checked for compatibility.
In the particular example, `C` does not get a type at first and therefore falls back to `Any`, which is (as of now) not compatible with integer.
There is two possible ways to resolve this for the general case:
1. Check for compatibility in the bodies before doing the fallback assignment to `Any`.
2. Do not throw an error on incompatible body types but just a warning (since then join will trivially be empty but nothing goes really wrong).

I would not go for 1 since our type inference might not cover all cases for values to be propagated to `C` in this case, e.g. if there is some variable assignment `?X = 42` in the body for example. So even though we cannot infer any type for `C`, I'm not quite sure that it will really remain empty.

I also would not go for 2 for now since I'm not sure how the physical layer would react in such a case. Anyhow, an empty join is probably not intended so throwing an error to the user is not the worst choice imho. The error message can be improved though.
Also, in the current efforts to change the API of the physical layer, the role of the type inference will most likely change and will also affect this particular case.

Long story short, this is intended behavior as of now.
